crates/ruff_linter/src/rules/ruff/rules/class_dict_annotations.rs

```diff
@@ -0,0 +1,133 @@
+///     above).
+///
+/// See [Python Annotations Best Practices](https://docs.python.org/3.14/howto/annotations.html)
+/// for alternatives.
+///
+/// ## Example
+///
+/// ```python
+/// cls.__dict__.get("__annotations__", {})
+/// ```
+///
+/// On Python 3.14+, use instead:
+/// ```python
+/// import annotationlib
+///
+/// annotationlib.get_annotations(cls)
+/// ```
+///
+/// On Python 3.10+, use instead:
+/// ```python
+/// import inspect
+///
+/// inspect.get_annotations(cls)
+/// ```
+///
+/// On Python < 3.10 with [typing-extensions](https://pypi.org/project/typing-extensions/)
+/// installed, use instead:
+/// ```python
+/// import typing_extensions
+///
+/// typing_extensions.get_annotations(cls)
+/// ```
+///
+/// ## Fix safety
+///
+/// No autofix is currently provided for this rule.
+///
+/// ## Fix availability
+///
+/// No autofix is currently provided for this rule.
+///
+/// ## References
+/// - [Python Annotations Best Practices](https://docs.python.org/3.14/howto/annotations.html)
+#[derive(ViolationMetadata)]
+pub(crate) struct ClassDictAnnotations;
+
+impl Violation for ClassDictAnnotations {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::None;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "Use `annotationlib.get_annotations` (Py3.14+), \
+        `inspect.get_annotations` (Py3.10+), or \
+        `typing_extensions.get_annotations` (Py<3.10 w/ `typing-extensions`) \
+        instead of `__dict__.get('__annotations__')`"
+            .to_string()
+    }
+}
+
+/// RUF061
+pub(crate) fn class_dict_annotations(checker: &Checker, call: &ExprCall) {
+    // Only apply this rule for Python 3.10 and newer unless `typing-extensions` is enabled.
+    if checker.target_version() < PythonVersion::PY310 && !checker.settings.typing_extensions {
+        return;
+    }
+
+    // Expected pattern: <identifier>.__dict__.get("__annotations__" [, <default>])
+    // Here, `call` is the `.get(...)` part.
+
+    // 1. Check that the `call.func` is `get`
+    let get_attribute = match call.func.as_ref() {
+        Expr::Attribute(attr) if attr.attr.as_str() == "get" => attr,
+        _ => return, // Not a call to an attribute named "get"
+    };
+
+    // 2. Check that the `get_attribute.value` is `__dict__`
+    match get_attribute.value.as_ref() {
+        Expr::Attribute(attr) if attr.attr.as_str() == "__dict__" => {}
+        _ => return, // The object of ".get" is not an attribute named "__dict__"
+    }
+
+    // At this point, we have `<identifier>.__dict__.get`.
+
+    // 3. Check arguments to `.get()`:
+    //    - No keyword arguments.
+    //    - One or two positional arguments.
+    //    - First positional argument must be the string literal "__annotations__".
+    //    - The value of the second positional argument (if present) does not affect the match.
+    if call.arguments.keywords.is_empty()
+        && (call.arguments.args.len() == 1 || call.arguments.args.len() == 2)
```

Comment:
nit: I'd probably convert this part into an early return: 
```suggestion
    if !call.arguments.keywords.is_empty() || call.arguments.len() > 2 {
        return;
    }
```